### PR TITLE
MA Authentication

### PR DESCRIPTION
**Pull request by dsamojlenko**
# What this does
- Adds basic authentication and policies for the MA routes (sessions/view/etc).
- Adds User model with checkPassword instanceMethod
- Adds [bcrypt](https://www.npmjs.com/package/bcrypt) for [password hashing](https://codahale.com/how-to-safely-store-a-password/)
- Adds isAuthenticated middleware
- Adds tests around auth functionality
- Adds command line utility for creating a user account
- Adds database seeder to create a user for testing
- Cleanup/simplify `config/policies.js`
- Adds FeatureFlags utility to enable feature flagging
- Updates Dockerfile to use multi-stage build

## To test
First, enable the Auth feature by setting `FEATURE_AUTH=true` in .env (you'll need to reload sails for the config to take effect). Then:
- attempt to access `/en/sessions`
- you should find yourself redirected to `/en/login`
- attempt to access `/en/sessions/1/view`
- you should find yourself redirected to `/en/login`
- run the seeder `npm run db:seed` to create a test user
- visit `/en/login` and use the credentials:
    - username: test@user.com
    - password: secret
- on successful login you should find yourself on the `/en/sessions` page, and have access to view sessions
- visit `/en/logout` and you should find yourself logged out and unable to access the sessions routes again
- try logging in with incorrect credentials, notice you will not get logged in

## Create user cli
Use the command line utility to create a user account:
```
sails run create-user --name="Full name" --email="name@email.com" --password="secret"
```

## Auth tests
Run the test suite by running `npm run cypress:open` and then select `Run all specs` or select the test spec to run.

Note that the Auth tests run `npm run db:seed` to ensure there is a known user for testing.

## Feature flags
This PR adds a Feature Flags utility. Feature Flags are just boolean flags specified in the environment, so to use, start by creating a feature flag in .env:

```
FEATURE_AUTH=true
```

Then, when you want to switch functionality on or off:

```
const featureFlags = require('FeatureFlags');

//

if (featureFlags.isEnabled('FEATURE_AUTH')) {
    // enable auth
}
```

## Dockerfile multi-stage build
This PR requires the `bcrypt` package for password encryption. Bcrypt needs to be built for Alpine on install, but Alpine does not have the tools to build from source. So changed our Dockerfile to a multi-stage build so we can install dependencies (python, gcc, make, etc) required for building bcrypt in an intermediate container, but then output a final image without those added dependencies.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,27 +1,29 @@
...
 RUN npm install
 
-# Copy in the rest of the files
-COPY . ./
+FROM node:lts-alpine3.11 AS final
+WORKDIR /app
+COPY --from=base /app /app
+
+# Install SailsJS
+RUN npm install sails@${SAILS_VERISON} -g
```
Comment (CalvinRodo) [inline Dockerfile]: Could we run this before the files are copied from the last container I'm pretty sure this will speed up the container build (admit I could be wrong here).

File: Dockerfile [hunk 2]
```diff
@@ -1,27 +1,29 @@
...
-RUN npm install sails@${SAILS_VERISON} -g
+# Need to install python to compile bcrypt dependency
+RUN apk --no-cache add --virtual native-deps \
+  g++ gcc libgcc libstdc++ linux-headers autoconf automake make nasm python git && \
+  npm install --quiet node-gyp -g
 
-# Copy in build files
-COPY package.json package-lock.json ./
+# Package files
+COPY . /app/
```
Comment (CalvinRodo) [inline Dockerfile]: Why did you remove copying over the package.json files? This will increase the build time significantly.

File: views/pages/login.njk [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+{% block content %}
+    <h1>{{ __('Login') }}</h1>
+
+<div class="flex justify-center"> 
+    <div class="w-full max-w-sm">
+        <form action="{{ route('login.post') }}" method="POST">
+            <input type="hidden" name="_csrf" value="{{ _csrf }}">
+
+            <label for="username">Email</label>
+            <input type="text" name="email" id="username" autofocus="true">
```
Comment (CalvinRodo) [inline views/pages/login.njk]: Why aren't we using Nunjucks?

File: api/models/User.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    email: {
+      type: Sequelize.STRING,
+      required: true,
+      unique: true
+    },
+    password: {
+      type: Sequelize.STRING,
+      required: true,
+      set(value) {
+        this.setDataValue('password', bcrypt.hashSync(value, 10));
```
Comment (CalvinRodo) [inline api/models/User.js]: Should we not be salting the password when we hash it?

File: Dockerfile (hunk 2 above)
Comment (dsamojlenko) [reply Dockerfile]: ... good question... I just pushed a change, see how it works.

File: api/models/User.js (hunk 4 above)
Comment (dsamojlenko) [reply api/models/User.js]: Apparently bcrypt work factors take care of this https://codahale.com/how-to-safely-store-a-password/ 

...doing some more research on this.

Comment (dsamojlenko) [reply api/models/User.js]: image

Comment (dsamojlenko) [reply api/models/User.js]: https://stackoverflow.com/questions/6832445/how-can-bcrypt-have-built-in-salts

File: views/pages/login.njk (hunk 3 above)
Comment (dsamojlenko) [reply views/pages/login.njk]: Good question. Now we are.

File: Dockerfile (hunk 1 above)
Comment (dsamojlenko) [reply Dockerfile]: moved it up

File: Dockerfile [hunk 5]
```diff
@@ -1,27 +1,31 @@
-FROM node:lts-alpine3.11
```
Comment (jeffmaher) [inline Dockerfile]: Dockerfile changes breaks using Docker for development. Added #153 as a PR into this one to fix that. Let me know what you think.
